lib/hipchat/errors.rb

```diff
@@ -1,14 +1,46 @@
 module HipChat
-  class ServiceError        < StandardError; end
-  class UnknownRoom         < ServiceError; end
-  class RoomNameTooLong     < ServiceError; end
-  class RoomMissingOwnerUserId < ServiceError; end
-  class Unauthorized        < ServiceError; end
-  class UsernameTooLong     < ServiceError; end
-  class UnknownResponseCode < ServiceError; end
-  class UnknownUser         < ServiceError; end
-  class InvalidApiVersion   < ServiceError; end
-  class InvalidUrl          < ServiceError; end
-  class InvalidEvent        < ServiceError; end
-  class UnknownWebhook      < ServiceError; end
+  class ServiceError            < StandardError;  end
+  class RoomNameTooLong         < ServiceError;   end
+  class RoomMissingOwnerUserId  < ServiceError;   end
+  class Unauthorized            < ServiceError;   end
+  class BadRequest              < ServiceError;   end
+  class MethodNotAllowed        < ServiceError;   end
+  class UsernameTooLong         < ServiceError;   end
+  class UnknownResponseCode     < ServiceError;   end
+  class InvalidApiVersion       < ServiceError;   end
+  class InvalidUrl              < ServiceError;   end
+  class InvalidEvent            < ServiceError;   end
+  class ObjectNotFound          < ServiceError;   end
+  class UnknownRoom             < ObjectNotFound; end
+  class UnknownUser             < ObjectNotFound; end
+  class UnknownWebhook          < ObjectNotFound; end
+  class TooManyRequests         < ServiceError;   end
+
+  class ErrorHandler
+
+    # Pass-through to catch error states and raise their appropriate exceptions
+    # @param klass {Symbol} The class of object we are handling the error for
+    # @param identifier {String} An identifying string for the object causing the error
+    # @param response {HTTParty::Response} The HTTParty response/request object
+    def self.catch_and_raise_exception_for(klass, identifier, response)
+      not_found_exception = Module.const_get(HipChat.to_s.to_sym).const_get("Unknown#{klass.capitalize}".to_sym)
```

Comment:
This is a little hard to decode exactly what's happening here and why. At least add some doco for this.